### ISSUE TITLE
Async changes in updating hook

### ISSUE THREAD

**Issue by cfilipov** [dfahlander/Dexie.js#372]
I am trying to figure out the best approach for using hooks to add some fields to an object when it gets stored or changed. 

The basic idea is there are `entry` objects which must contain a bunch of properties that are based on some complex queries and calculations of other entries. These calculated properties are all stored under a property called `derived`. It would far prohibitively expensive to calculate `entry.derived` every time it is needed, or when it is read from the DB. Instead I opted to populate the `derived` property ahead of time, and hooks seem to be the best place to do this.

This seems to be no problem for the `creating` hook. However, I also need to re-generate `derived` if any other property in `entry` is changed. The `updating` hook requires I submit additional changes by returning them, which is problematic because the only way I can generate the change is via an async call.

Below is some minimal code that tries to demonstrate the problem. I haven't tried option `B` yet, but I suspect it won't work either.

```
const entryDerivedData = function(entry) {
    // query a bunch of data from entries table then do some calculations
    return db.entries.where('exerciseID').equals(entry.exerciseID).toArray()
        .then(e => {
            // do some calculation and return
            return calculateDerivedData(e);
        });
};

// A: Can't do this because `hook` isn't expecting a generator func
db.entries.hook('updating', function*(mods, primKey, entry, transaction) {
    const derived = yield entryDerivedData(entry);
    return derived;
});

// B: Another possibility, but probably won't work either
db.entries.hook('updating', function(mods, primKey, entry, transaction) {
    transaction.scopeFunc = function() {
        return entryDerivedData(entry)
            .then(derived => {
                // Won't this result in another invocation of the updating hook?
                return db.entries.update(entry.id, {derived});
            });
    };
});
```

**Comment by dfahlander** [comment]
Unfortunately, hooks are synchronous and there's no way to do async calls within them. This is something that is going to change but I cannot promise when. Hoping to rewrite the hook framework within the next 6 months or so and allow bulk hooks (more performant) that can be async (keeping backward compatibility for existing hooks).

Until then, you could utilize the fact that the hook is always called within a transaction (no matter whether the user does an explicit transaction or not), and you can do additional operations onto the current transaction. Just need to make sure you don't end up in an infinite loop as your additional changes may trigger your hook again.

An example would be this:

```js
db.entries.hook('creating', (primKey, entry, trans) => {
  entryDerivedData(entry).then(derived => {
        db.entries.update(primKey, { derived }).catch (e => {
            // Failed to update. Abort transaction or rethrow error:
           throw new Error ("Could not make sure derived property was set properly");
        });
    });
});

db.entries.hook('updating', (mods, primKey, entry, trans) => {
    if ('derived' in mods) return; // We're the one triggering this change. Ignore.
    // First, apply the mods onto entry:
    var entryClone = Dexie.deepClone(entry);
    Object.keys(mods).forEach(keyPath => {
        if (mods[keyPath] === undefined)
            Dexie.delByKeyPath(entryClone, keyPath);
        else
            Dexie.setByKeyPath(entryClone, keyPath, mods[keyPath]);
    });
    
    entryDerivedData(entryClone).then(derived => {
        db.entries.update(primKey, { derived }).catch (e => {
            // Failed to update. Abort transaction or rethrow error:
           throw new Error ("Could not make sure derived property was set properly");
        });
    });
});


```

**Comment by dfahlander** [comment]
If you really need to enforce that the user cannot change "derived" property without enforcing your policy, you would need to assure that the hook caller is yourself. This is possible to accomplish via zones / Dexie.PSD:

```js

const secretProp = Math.random().toString();

db.entries.hook('updating', (mods, primKey, entry, trans) => {
    if (Dexie.Promise.PSD[secretProp]) return; // We're the one triggering this change. Ignore.
  
    ...
    
    Dexie.Promise.newPSD(()=>{
        // You enter a new zone that will be derived for all async calls
        // originating from this scope. Mark the zone with a private property:
        Dexie.Promise.PSD[secretProp] = true;

        entryDerivedData(entryClone).then(derived => {
            db.entries.update(primKey, { derived }).catch (e => {
               // Failed to update. Abort transaction...
               throw new Error ("Could not make sure derived property was set properly");
            });
        });
    });
});
```

**Comment by cfilipov** [comment]
Thank you! I did not think of checking for `derived` in the mods object, now that you've shown me it looks like an obvious solution (until async hooks are ready).

By the way, why is this necessary?

    // First, apply the mods onto entry:
    var entryClone = Dexie.deepClone(entry);
    Object.keys(mods).forEach(keyPath => {
        if (mods[keyPath] === undefined)
            Dexie.delByKeyPath(entryClone, keyPath);
        else
            Dexie.setByKeyPath(entryClone, keyPath, mods[keyPath]);
    });

Isn't this going to happen already anyway after the hook returns?

**Comment by dfahlander** [comment]
Yes it's going to happen anyway, but in case your code (entryDerivedData) depends on the updated property, like excerciseId, or would expect an object where the mods has been applied

**Comment by cfilipov** [comment]
Sorry, one more follow-up to this..

How are hooks and schema upgrades meant to work together? What's happening is that I implemented the hooks as described above (thanks again!), however, the hooks are of course written with the *current* schema in mind, but it appears that the upgrade function is called before the schema is fully migrated.

Basically, the `entryDerivedData` function depends on a new store/table that was added as part of schema version 2 and it's getting called in the hook when the migration is taking place.

```
db.version(0.2)
	.stores({ entries: ..., exercises: ... })
	.upgrade(transaction => {
		return transaction.entries.orderBy('start').modify(function(entry) {
			// Replace old entry with slightly different one (`derived` object is wiped out)
			// This results in a call to one of the hooks which tries to use the `exercises` table that doesn't exist yet.
			this.value = {
				id: entry.id,
				...
			};
		})
		.catch(function(err) {
			throw new Error(`Failed to migrate schema. ${err}`);
		});
	});
db.version(0.1).stores({entries: '++id,start,xid'});
```

I suppose I could use `Dexie.PSD` to prevent the hook from getting called, then somehow manually trigger the re-generation of `derived` on every entry after the migration is done. Is there a better way?

**Comment by dfahlander** [comment]
Hooks are not meant to run during upgrades. If they do, it's a bug. If a developer changes or adds hooks between versions, it's up to the developer to add an upgrader that transforms the data to the new format. That's how it "should" work, but...

The upgrade framework is in need of a rewrite (#105, #287) that was initiated in pull request #293 but have gone stall. It's my plan to go through it and make sure it works as indended. Especially, issues like the one you recently got, happens when the "installed" db is not in sync with the declared db, which can happen of many reasons.

Any inputs of how you would like it to work, are welcome, so I could take those into consideration when rewriting it.

**Comment by cfilipov** [comment]
> Especially, issues like the one you recently got, happens when the "installed" db is not in sync with the declared db, which can happen of many reasons.

This might be what happened for me. I recall getting some exceptions during an upgrade while developing, and instead of starting from a clean environment I kept working from there. I haven't tried to reproduce this yet but I suspect the hook will not fire (as expected) once I do this cleanly.

> it's up to the developer to add an upgrader that transforms the data to the new format. That's how it "should" work, but...

This is tricky for me.. because I am adding a new table which will be used in the hook once the upgrade is complete. To make things more complicated, the new table is populated from JSON (using the method described [here](https://github.com/dfahlander/Dexie.js/wiki/Dexie.on.populate)) so I can't use the underlying data during the upgrade function because it can't be async.

Is there a hook for when a schema upgrade has completed? The only solution I can see is possibly having some kind of `fillDerivedFields()` function that gets called after each schema upgrade. This ensures all tables required for filling derived fields exist and hooks are available. I don't see this capability documented anywhere though.

**Comment by cfilipov** [comment]
Update on some topics brought up this issue: the error I was getting while migrating was actually due to #473. It's also worth noting that I think my initial issue was approaching the problem incorrectly in the first place. Having the entry derived data on the entry itself made things more complicated in the hook because it required some mechanism to prevent the hook from triggering itself (see Dexie.PSD and other suggestion from @dfahlander above). 

The ultimate solution was to normalize the data. Have the derived data of the entry live in its own table means that hooks on the entries tables can write to the deriveddata table and not worry about triggering itself.

Additionally, using Dexie.Observable instead of hooks means I don't have to manually apply the mods in the updating hook since the observable happens after the data is committed.

With this, I am no longer in need of async hook, but they are likely still useful to have.
